Mathlib/Data/Fin/Tuple/Basic.lean

```diff
@@ -0,0 +1,878 @@
+#align fin.cons_zero Fin.cons_zero
+
+/-- Updating a tuple and adding an element at the beginning commute. -/
+@[simp]
+theorem cons_update : cons x (update p i y) = update (cons x p) i.succ y :=
+  by
+  ext j
+  by_cases h : j = 0
+  · rw [h]
+    simp [Ne.symm (succ_ne_zero i)]
+  · let j' := pred j h
+    have : j'.succ = j := succ_pred j h
+    rw [← this, cons_succ]
+    by_cases h' : j' = i
+    · rw [h']
+      simp
+    · have : j'.succ ≠ i.succ := by rwa [Ne.def, succ_inj]
+      rw [update_noteq h', update_noteq this, cons_succ]
+#align fin.cons_update Fin.cons_update
+
+/-- As a binary function, `Fin.cons` is injective. -/
+theorem cons_injective2 : Function.Injective2 (@cons n α) := fun x₀ y₀ x y h ↦
+  ⟨congr_fun h 0, funext fun i ↦ by simpa using congr_fun h (Fin.succ i)⟩
+#align fin.cons_injective2 Fin.cons_injective2
+
+@[simp]
+theorem cons_eq_cons {x₀ y₀ : α 0} {x y : ∀ i : Fin n, α i.succ} :
+    cons x₀ x = cons y₀ y ↔ x₀ = y₀ ∧ x = y :=
+  cons_injective2.eq_iff
+#align fin.cons_eq_cons Fin.cons_eq_cons
+
+theorem cons_left_injective (x : ∀ i : Fin n, α i.succ) : Function.Injective fun x₀ ↦ cons x₀ x :=
+  cons_injective2.left _
+#align fin.cons_left_injective Fin.cons_left_injective
+
+theorem cons_right_injective (x₀ : α 0) : Function.Injective (cons x₀) :=
+  cons_injective2.right _
+#align fin.cons_right_injective Fin.cons_right_injective
+
+/-- Adding an element at the beginning of a tuple and then updating it amounts to adding it
+directly. -/
+theorem update_cons_zero : update (cons x p) 0 z = cons z p := by
+  ext j
+  by_cases h : j = 0
+  · rw [h]
+    simp
+  · simp only [h, update_noteq, Ne.def, not_false_iff]
+    let j' := pred j h
+    have : j'.succ = j := succ_pred j h
+    rw [← this, cons_succ, cons_succ]
+#align fin.update_cons_zero Fin.update_cons_zero
+
+/-- Concatenating the first element of a tuple with its tail gives back the original tuple -/
+@[simp]
+theorem cons_self_tail : cons (q 0) (tail q) = q := by
+  ext j
+  by_cases h : j = 0
+  · rw [h]
+    simp
+  · let j' := pred j h
+    have : j'.succ = j := succ_pred j h
+    rw [← this]
+    unfold tail
+    rw [cons_succ]
+#align fin.cons_self_tail Fin.cons_self_tail
+
+-- Porting note: Mathport removes `_root_`?
+/-- Recurse on an `n+1`-tuple by splitting it into a single element and an `n`-tuple. -/
+@[elab_as_elim]
+def consCases {P : (∀ i : Fin n.succ, α i) → Sort v} (h : ∀ x₀ x, P (Fin.cons x₀ x))
+    (x : ∀ i : Fin n.succ, α i) : P x :=
+  _root_.cast (by rw [cons_self_tail]) <| h (x 0) (tail x)
+#align fin.cons_cases Fin.consCases
+
+@[simp]
+theorem cons_cases_cons {P : (∀ i : Fin n.succ, α i) → Sort v} (h : ∀ x₀ x, P (Fin.cons x₀ x))
+    (x₀ : α 0) (x : ∀ i : Fin n, α i.succ) : @consCases _ _ _ h (cons x₀ x) = h x₀ x := by
+  rw [consCases, cast_eq]
+  congr
+#align fin.cons_cases_cons Fin.cons_cases_cons
+
+-- Porting note: I am not perfectly confident in my choice to add `h0 h` to `consInduction` at end
```

Comment:
If you mean the extra arguments in the recursive call, that seems standard in lean4
```suggestion
```